bblocks/import_tools/world_bank_projects.py

```diff
@@ -0,0 +1,562 @@
+        theme_level (int): The current level of theme nesting.
+    """
+    # Append the current theme to the list
+    _append_theme_to_list(
+        proj_id=proj_id, theme_list=theme_list, theme_names=theme_names, theme=theme
+    )
+
+    # Recursively call this function for each nested theme
+    nested_theme_key = f"theme{theme_level + 1}"
+    for nested_theme in theme.get(nested_theme_key, []):
+        _parse_themes(
+            proj_id=proj_id,
+            theme_list=theme_list,
+            theme_names=theme_names + [nested_theme["name"]],
+            theme=nested_theme,
+            theme_level=theme_level + 1,
+        )
+
+
+def clean_theme(data: dict) -> list[dict] | list:
+    """Clean theme data from a nested list to a list of dictionaries with theme names and
+    percentages.
+    If there are no themes, an empty list will be returned.
+
+    Args:
+        data: data from API
+
+    Returns:
+        list of dictionaries with theme names and percentages
+    """
+
+    # if there are no themes, return an empty list
+    if "theme_list" not in data:
+        # return [{'project ID': proj_id}]
+        return []
+
+    theme_list = []
+    proj_id = data["id"]
+    for theme1 in data["theme_list"]:
+        _parse_themes(proj_id, theme_list, [theme1["name"]], theme1, 1)
+
+    return theme_list
+
+
+def _get_sector_data(d: dict) -> dict:
+    """Get sector percentages from a project dictionary
+
+    the function first finds all available sectors
+    It then finds all  fields from the json starting with 'sector' and ending with a number
+    and gets a dictionary of the sector name and percentage
+    If there are any sectors missing from the dict and the total percentage is less than 100
+    the missing sector is added with the remaining percentage.
+    If the total is still not 100, it will raise an error to indicate a
+    problem with the data.
+
+    args:
+        d: project dictionary
+    """
+
+    sectors_dict = {}  # empty dict to store sector data as {sector_name: percent}
+    sector_names = [v["Name"] for v in d["sector"]]  # get list of sector names
+
+    # get sectors fields which should contain percentages
+    sectors = {key: value for key, value in d.items() if re.search(r"^sector\d+$", key)}
+
+    # get available sector percentages
+    for _, v in sectors.items():
+        if isinstance(v, dict):
+            sectors_dict[v["Name"]] = v["Percent"]
+
+    # check if there are missing sectors from the dict
+    if (len(sectors_dict) == len(sectors) - 1) and (sum(sectors_dict.values()) < 100):
+        # loop through all the available sectors
+        for s in sector_names:
+            # if a sectors has not been picked up it must be the missing sector
+            if s not in sectors_dict:
+                sectors_dict[s] = 100 - sum(sectors_dict.values())
+
+    if sum(sectors_dict.values()) != 100:
+        raise ValueError("Sector percentages don't add up to 100%")
```

Comment:
@jm-rivera we need to revaluate the strategy here. In some instances the sectors that exist in the list `sectors` they do exist in the disctionary `sectors_dict`. However, the missing sector does not exist in the website or the excel download from the main projects page. see for example project `P178202` - "waste management is not included in the [project page](https://projects.worldbank.org/en/projects-operations/project-detail/P178202). In the api response, we don't see a percentage - this is the response value `'sector5': 'Waste Management!$!11!$!WB'`.

My original assumption was that if sectors are specified their percentage allocations should add to 100%. However I have noticed this is not always the case. Project `P073479` has 1 sector allocated for ICT technologies at 24%. If sector data does not have to add up to 100% there is no way to determine the percent that should be allocated to those missing sectors.

I'd suggest then that we remove this calculation and just parse the sectors which have a percent allocation
Open to suggestions.